File: src/main/java/eu/qwsome/sql/Select.java
package eu.qwsome.sql;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import eu.qwsome.sql.TableJoin.Type;
import eu.qwsome.sql.condition.Condition;
import eu.qwsome.sql.condition.ValueConstructor;

/**
 * This class simplifies dynamic sql generation.
 *
 * @author Lukáš Kvídera
 */
public class Select implements Query {

  /**
   * List of columns that will be selected.
   */
  private final List<String> columns = new ArrayList<>();
  /**
   * A list of joins that generates multiple JOIN clauses.
   */
  private final List<Join> joins = new ArrayList<>();
  /**
   * A list of conditions from subselects
   */
  private final List<Condition> subselectConditions = new ArrayList<>();
  /**
   * True if SELECT DISTINCT shall be used
   */
  private boolean distinct;
  /**
   * Datasource after FROM (might be table or sub-query).
   */
  private String source;
  /**
   * Condition used in WHERE clause.
   */
  private Condition condition;
  private OrderBy orderBy;
  private GroupBy groupBy;
  /**
   * This may contain conditions from unin used as subselect
   */
  private final List<ValueBinding> union = new ArrayList<>();


  /**
   * Creates select with single column.
   *
   * @param column name of column to be selected
   */
  private Select(final String column) {
    this.columns.add(column);
  }


  /**
   * Creates select with multiple columns.
   *
   * @param columns names of columns to be selected
   */
  private Select(final String[] columns) {
    this.columns.addAll(Arrays.asList(columns));
  }


  /**
   * This mehod creates a select instance that can be effectively used as following snippet:
   * <p>
   * {@code
   * select().from("table").toSql()
   * }
   * <p>
   * that is translated into:
   * <p>
   * {@code
   * SELECT * FROM table
   * }
   *
   * @return select builder
   */
  public static Select select() {
    return new Select("*");
  }


  /**
   * This mehod creates a select instance that can be effectively used as following snippet:
   * <p>
   * {@code
   * select("column").from("table").toSql()
   * }
   * <p>
   * that is translated into:
   * <p>
   * {@code
   * SELECT column FROM table
   * }
   *
   * @param column name that will be selected
   *
   * @return select builder
   */
  public static Select select(final String column) {
    return new Select(column);
  }


  /**
   * This mehod creates a select instance that can be effectively used as following snippet:
   * <p>
   * {@code
   * select("column1", "column2", ...).from("table").toSql()
   * }
   * <p>
   * that is translated into:
   * <p>
   * {@code
   * SELECT column1, column2, ... FROM table
   * }
   *
   * @return select builder
   */
  public static Select select(final String... columns) {
    return new Select(columns);
  }


  /**
   * This method enables use of distinction.
   *
   * @return select with enabled distinction
   */
  public Select distinct() {
    this.distinct = true;
    return this;
  }


  /**
   * This method sets a source table for select.
   *
   * @param table to be used in FROM clause
   *
   * @return next phase that allows only relevant methods
   */
  public TableSelectedPhase from(final String table) {
    this.source = table;
    return new TableSelectedPhase();
  }


  /**
   * This method sets a source source sub-query for select.
   *
   * @param subquery source sub-query
   * @param alias alias of the sub-query
   *
   * @return next selection phase
   */
  public TableSelectedPhase from(final Query subquery, final String alias) {
    this.source = "( "
        .concat(subquery.toSql())
        .concat(" ) AS ")
        .concat(alias);

    if (subquery instanceof ConditionsBuiltPhase) {
      this.subselectConditions.add(((ConditionsBuiltPhase) subquery).getSelect().condition);
    } else if (subquery instanceof ValueBinding) {
      this.union.add((ValueBinding) subquery);
    }

    return new TableSelectedPhase();
  }


  /**
   * Generates final SQL.
   *
   * @return generated SQL
   */
  @Override
  public String toSql() {
    final StringBuilder builder = new StringBuilder();
    builder.append("SELECT ")
        .append(this.distinct ? "DISTINCT " : "")
        .append(getColumns())
        .append(" FROM ")
        .append(this.source)
    ;

    if (!this.joins.isEmpty()) {
      for (final var join : this.joins) {
        join.appendTo(builder);
      }
    }

    if (this.condition != null) {
      builder.append(" WHERE ");
      this.condition.appendTo(builder);
    }

    if (this.groupBy != null) {
      this.groupBy.appendTo(builder);
    }

    if (this.orderBy != null) {
      this.orderBy.appendTo(builder);
    }


    return builder.toString();
  }


  /**
   * Adds an order by clause to the statement.
   *
   * @param columns that defines ordering
   *
   * @return order by phase
   */
  private OrderByPhase orderBy(final Column... columns) {
    this.orderBy = new OrderBy(columns);
    return new OrderByPhase();
  }


  private GroupByPhase groupBy(final Column... groupByColumns) {
    this.groupBy = new GroupBy(groupByColumns);
    return new GroupByPhase();
  }


  private ValueConstructor toValuesInternal() {
    final var values = new ValueConstructor();

    Select.this.joins.stream()
        .map(Join::toValues)
        .forEach(values::add)
    ;

    Select.this.subselectConditions.stream()
        .map(Condition::getValues)
        .forEach(values::add)
    ;

    Select.this.union.stream()
        .map(ValueBinding::toValues)
        .forEach(values::add)
    ;

    if (this.condition != null) {
      values.add(Select.this.condition.getValues());
    }

    return values;
  }


  /**
   * This method generates a list of columns concatenated with comma.
   *
   * @return list of columns usable in SQL
   */
  private String getColumns() {
    return String.join(", ", this.columns);
  }


  public class TableSelectedPhase implements ValueBinding {
    /**
     * @see Select#toSql()
     */
    @Override
    public String toSql() {
      return Select.this.toSql();
    }


    /**
     * This method allows creation of where clause.
     *
     * @param condition condition used to filter data
     *
     * @return next phase that allows only relevant methods
     */
    public ConditionsBuiltPhase where(final Condition condition) {
      Select.this.condition = condition;
      return new ConditionsBuiltPhase();
    }


    /**
     * Adds an order by clause to the statement.
     *
     * @param columns that defines ordering
     *
     * @return order by phase
     */
    public OrderByPhase orderBy(final Column... columns) {
      return Select.this.orderBy(columns);
    }


    /**
     * Adds a join clause to the statement.
     *
     * @param joinTable table to be joined
     *
     * @return next phase that allows only relevant methods
     */
    public TableJoinPhase join(final String joinTable) {
      return new TableJoinPhase(joinTable, Type.INNER);
    }


    /**
     * Adds a join clause to the statement.
     *
     * @param subselect to be joined
     *      * @param alias for subselect
     *
     * @return next phase that allows only relevant methods
     */
    public SubselectJoinPhase join(final SubselectValueHolder subselect, final String alias) {
      return new SubselectJoinPhase(subselect, alias, Type.INNER);
    }


    /**
     * Adds a left join clause to the statement.
     *
     * @param joinTable table to be joined
     *
     * @return next phase that allows only relevant methods
     */
    public TableJoinPhase leftJoin(final String joinTable) {
      return new TableJoinPhase(joinTable, Type.LEFT);
    }


    /**
     * Adds a join clause to the statement.
     *
     * @param subselect to be joined
     * @param alias for subselect
     *
     * @return next phase that allows only relevant methods
     */
    public SubselectJoinPhase leftJoin(final SubselectValueHolder subselect, final String alias) {
      return new SubselectJoinPhase(subselect, alias, Type.LEFT);
    }


    @Override
    public ValueConstructor toValues() {
      return toValuesInternal();
    }


    public GroupByPhase groupBy(final Column... groupByColumns) {
      return Select.this.groupBy(groupByColumns);
    }
  }

  /**
   * Final phase.
   */
  public class ConditionsBuiltPhase implements ValueBinding {
    /**
     * @return generated SQL
     *
     * @see Select#toSql()
     */
    @Override
    public String toSql() {
      return Select.this.toSql();
    }


    @Override
    public ValueConstructor toValues() {
      return toValuesInternal();
    }


    Select getSelect() {
      return Select.this;
    }


    /**
     * Adds an order by clause to the statement.
     *
     * @param columns that defines ordering
     *
     * @return order by phase
     */
    public OrderByPhase orderBy(final Column... columns) {
      return Select.this.orderBy(columns);
    }


    public GroupByPhase groupBy(final Column... columns) {
      return Select.this.groupBy(columns);
    }


    public Union union(final ConditionsBuiltPhase otherSelect) {
      return Union.of(this, otherSelect);
    }
  }

  /**
   * Phase available after {@link TableSelectedPhase}
   */
  public class TableJoinPhase {

    /**
     * Table that will be joined.
     */
    private final String joinTable;
    private final Type type;


    private TableJoinPhase(final String joinTable, final Type type) {
      this.joinTable = joinTable;
      this.type = type;
    }


    /**
     * Creates a join clues with condition.
     *
     * @param condition that will be used in ON clause
     *
     * @return next phase that allows only relevant methods
     */
    public TableSelectedPhase on(final Condition condition) {
      switch (this.type) {
        case INNER:
          Select.this.joins.add(new InnerTableJoin(this.joinTable, condition));
          break;
        case LEFT:
          Select.this.joins.add(new LeftTableJoin(this.joinTable, condition));
          break;
        default:
          throw new UnsupportedOperationException(this.type.toString());
      }
      return new TableSelectedPhase();
    }
  }

  /**
   * Phase available after {@link TableSelectedPhase}
   */
  public class SubselectJoinPhase {

    /**
     * Table that will be joined.
     */
    private final SubselectValueHolder subselect;
    private final String alias;
    private final Type type;


    private SubselectJoinPhase(final SubselectValueHolder subselect, String alias, final Type type) {
      this.subselect = subselect;
      this.alias = alias;
      this.type = type;
    }


    /**
     * Creates a join clues with condition.
     *
     * @param condition that will be used in ON clause
     *
     * @return next phase that allows only relevant methods
     */
    public TableSelectedPhase on(final Condition condition) {
      switch (this.type) {
        case INNER:
          Select.this.joins.add(new SubselectInnerJoin(this.subselect, this.alias, condition));
          break;
        case LEFT:
          Select.this.joins.add(new SubselectLeftJoin(this.subselect, this.alias, condition));
          break;
        default:
          throw new UnsupportedOperationException(this.type.toString());
      }
      return new TableSelectedPhase();
    }
  }

  public class OrderByPhase implements ValueBinding {
    /**
     * @return generated SQL
     */
    @Override
    public String toSql() {
      return Select.this.toSql();
    }


    @Override
    public ValueConstructor toValues() {
      return toValuesInternal();
    }


  }

  public class GroupByPhase implements ValueBinding {
    /**
     * @return generated SQL
     */
    @Override
    public String toSql() {
      return Select.this.toSql();
    }


    @Override
    public ValueConstructor toValues() {
      return toValuesInternal();
    }


    public OrderByPhase orderBy(final Column... orderByColumns) {
      return Select.this.orderBy(orderByColumns);
    }
  }

}

File: src/main/kotlin/eu/qwsome/sql/Select.kt
/*
 * This class simplifies dynamic sql generation.
 *
 * @author Lukáš Kvídera
 */
package eu.qwsome.sql

import eu.qwsome.sql.condition.Condition
import eu.qwsome.sql.condition.ValueConstructor

class Select private constructor(vararg columns: String) : Query {

    /**
     * List of columns that will be selected.
     */
    private val columns: MutableList<String> = columns.toMutableList()

    /**
     * A list of joins that generates multiple JOIN clauses.
     */
    private val joins: MutableList<Join> = mutableListOf()

    /**
     * A list of conditions from subselects
     */
    private val subselectConditions: MutableList<Condition> = mutableListOf()

    /**
     * True if SELECT DISTINCT shall be used
     */
    private var distinct: Boolean = false

    /**
     * Datasource after FROM (might be table or sub-query).
     */
    private var source: String? = null

    /**
     * Condition used in WHERE clause.
     */
    private var condition: Condition? = null

    private var orderBy: OrderBy? = null
    private var groupBy: GroupBy? = null

    /**
     * This may contain conditions from union used as subselect
     */
    private val union: MutableList<ValueBinding> = mutableListOf()

    /**
     * This method enables use of distinction.
     *
     * @return select with enabled distinction
     */
    fun distinct(): Select = apply {
        distinct = true
    }

    /**
     * This method sets a source table for select.
     *
     * @param table to be used in FROM clause
     * @return next phase that allows only relevant methods
     */
    fun from(table: String): TableSelectedPhase {
        source = table
        return TableSelectedPhase()
    }

    /**
     * This method sets a source source sub-query for select.
     *
     * @param subquery source sub-query
     * @param alias alias of the sub-query
     * @return next selection phase
     */
    fun from(subquery: Query, alias: String): TableSelectedPhase {
        source = "( ${subquery.toSql()} ) AS $alias"

        when (subquery) {
            is ConditionsBuiltPhase -> {
                subquery.getSelect().condition?.let { subselectConditions.add(it) }
            }

            is ValueBinding -> {
                union.add(subquery)
            }
        }

        return TableSelectedPhase()
    }

    /**
     * Generates final SQL.
     *
     * @return generated SQL
     */
    override fun toSql(): String = buildString {
        append("SELECT ")
        if (distinct) {
            append("DISTINCT ")
        }
        append(getColumns())
        append(" FROM ")
        append(source)

        joins.forEach { join ->
            join.appendTo(this)
        }

        condition?.let {
            append(" WHERE ")
            it.appendTo(this)
        }

        groupBy?.appendTo(this)
        orderBy?.appendTo(this)
    }

    /**
     * Adds an order by clause to the statement.
     *
     * @param columns that defines ordering
     * @return order by phase
     */
    private fun orderBy(vararg columns: Column): OrderByPhase {
        orderBy = OrderBy(*columns)
        return OrderByPhase()
    }

    private fun groupBy(vararg groupByColumns: Column): GroupByPhase {
        groupBy = GroupBy(*groupByColumns)
        return GroupByPhase()
    }

    private fun toValuesInternal(): ValueConstructor {
        val values = ValueConstructor()

        joins.forEach { values.add(it.toValues()) }
        subselectConditions.forEach { values.add(it.getValues()) }
        union.forEach { values.add(it.toValues()) }

        condition?.let { values.add(it.getValues()) }

        return values
    }

    /**
     * This method generates a list of columns concatenated with comma.
     *
     * @return list of columns usable in SQL
     */
    private fun getColumns(): String {
        return columns.joinToString(", ")
    }

    inner class TableSelectedPhase : ValueBinding {
        /**
         * @see Select.toSql
         */
        override fun toSql(): String {
            return this@Select.toSql()
        }

        /**
         * This method allows creation of where clause.
         *
         * @param condition condition used to filter data
         * @return next phase that allows only relevant methods
         */
        fun where(condition: Condition?): ConditionsBuiltPhase {
            this@Select.condition = condition
            return ConditionsBuiltPhase()
        }

        /**
         * Adds an order by clause to the statement.
         *
         * @param columns that defines ordering
         * @return order by phase
         */
        fun orderBy(vararg columns: Column): OrderByPhase {
            return this@Select.orderBy(*columns)
        }

        /**
         * Adds a join clause to the statement.
         *
         * @param joinTable table to be joined
         * @return next phase that allows only relevant methods
         */
        fun join(joinTable: String): TableJoinPhase {
            return TableJoinPhase(joinTable, TableJoin.Type.INNER)
        }

        /**
         * Adds a join clause to the statement.
         *
         * @param subselect to be joined
         * @param alias for subselect
         * @return next phase that allows only relevant methods
         */
        fun join(subselect: SubselectValueHolder, alias: String): SubselectJoinPhase {
            return SubselectJoinPhase(subselect, alias, SubselectJoin.Type.INNER)
        }

        /**
         * Adds a left join clause to the statement.
         *
         * @param joinTable table to be joined
         * @return next phase that allows only relevant methods
         */
        fun leftJoin(joinTable: String): TableJoinPhase {
            return TableJoinPhase(joinTable, TableJoin.Type.LEFT)
        }

        /**
         * Adds a join clause to the statement.
         *
         * @param subselect to be joined
         * @param alias for subselect
         * @return next phase that allows only relevant methods
         */
        fun leftJoin(subselect: SubselectValueHolder, alias: String): SubselectJoinPhase {
            return SubselectJoinPhase(subselect, alias, SubselectJoin.Type.LEFT)
        }

        override fun toValues(): ValueConstructor {
            return toValuesInternal()
        }

        fun groupBy(vararg groupByColumns: Column): GroupByPhase {
            return this@Select.groupBy(*groupByColumns)
        }
    }

    /**
     * Final phase.
     */
    inner class ConditionsBuiltPhase : ValueBinding {
        /**
         * @return generated SQL
         * @see Select.toSql
         */
        override fun toSql(): String {
            return this@Select.toSql()
        }

        override fun toValues(): ValueConstructor {
            return toValuesInternal()
        }

        internal fun getSelect(): Select {
            return this@Select
        }

        /**
         * Adds an order by clause to the statement.
         *
         * @param columns that defines ordering
         * @return order by phase
         */
        fun orderBy(vararg columns: Column): OrderByPhase {
            return this@Select.orderBy(*columns)
        }

        fun groupBy(vararg columns: Column): GroupByPhase {
            return this@Select.groupBy(*columns)
        }

        fun union(otherSelect: ConditionsBuiltPhase): Union {
            return Union.of(this, otherSelect)
        }
    }

    /**
     * Phase available after [TableSelectedPhase]
     */
    inner class TableJoinPhase(
        private val joinTable: String,
        private val type: TableJoin.Type,
    ) {

        /**
         * Creates a join clause with condition.
         *
         * @param condition that will be used in ON clause
         * @return next phase that allows only relevant methods
         */
        fun on(condition: Condition): TableSelectedPhase {
            val join = when (type) {
                TableJoin.Type.INNER -> InnerTableJoin(joinTable, condition)
                TableJoin.Type.LEFT -> LeftTableJoin(joinTable, condition)
            }
            this@Select.joins.add(join)
            return TableSelectedPhase()
        }
    }

    /**
     * Phase available after [TableSelectedPhase]
     */
    inner class SubselectJoinPhase(
        private val subselect: SubselectValueHolder,
        private val alias: String,
        private val type: SubselectJoin.Type,
    ) {

        /**
         * Creates a join clause with condition.
         *
         * @param condition that will be used in ON clause
         * @return next phase that allows only relevant methods
         */
        fun on(condition: Condition): TableSelectedPhase {
            val join = when (type) {
                SubselectJoin.Type.INNER -> SubselectInnerJoin(subselect, alias, condition)
                SubselectJoin.Type.LEFT -> SubselectLeftJoin(subselect, alias, condition)
            }
            this@Select.joins.add(join)
            return TableSelectedPhase()
        }
    }

    inner class OrderByPhase : ValueBinding {
        /**
         * @return generated SQL
         */
        override fun toSql(): String {
            return this@Select.toSql()
        }

        override fun toValues(): ValueConstructor {
            return toValuesInternal()
        }
    }

    inner class GroupByPhase : ValueBinding {
        /**
         * @return generated SQL
         */
        override fun toSql(): String {
            return this@Select.toSql()
        }

        override fun toValues(): ValueConstructor {
            return toValuesInternal()
        }

        fun orderBy(vararg orderByColumns: Column): OrderByPhase {
            return this@Select.orderBy(*orderByColumns)
        }
    }

    companion object {
        /**
         * This method creates a select instance that can be effectively used as following snippet:
         *
         * ```
         * select().from("table").toSql()
         * ```
         *
         * that is translated into:
         *
         * ```
         * SELECT * FROM table
         * ```
         *
         * @return select builder
         */
        @JvmStatic
        fun select() = Select("*")

        /**
         * This method creates a select instance that can be effectively used as following snippet:
         *
         * ```
         * select("column").from("table").toSql()
         * ```
         *
         * that is translated into:
         *
         * ```
         * SELECT column FROM table
         * ```
         *
         * @param column name that will be selected
         * @return select builder
         */
        @JvmStatic
        fun select(column: String) = Select(column)

        /**
         * This method creates a select instance that can be effectively used as following snippet:
         *
         * ```
         * select("column1", "column2", ...).from("table").toSql()
         * ```
         *
         * that is translated into:
         *
         * ```
         * SELECT column1, column2, ... FROM table
         * ```
         *
         * @return select builder
         */
        @JvmStatic
        fun select(vararg columns: String): Select {
            return Select(*columns)
        }
    }
}
